### PR TITLE
Function to estimate wind speed at different heights

### LINKED ISSUE #2118: Expression for calculating wind speed at different heights
Most times, the wind speed provided in weather files is measured at a height of 10 m. However, in many cases it is required to estimate the wind speed at panel height. Which equation should be used in order to do this?

I found this issue (#1814), which mentions an equation from a Sandia report (I couldn't find the report, though... some help from Sandia people? @cwhanse @kandersolar). This is what is mentioned:
`Estimate wind speed at module height - use technique developed by menicucci and hall (sand84-2530) */`
`windmd=ws2*pow(height/9.144,0.2) + 0.0001`

And then there is the expression used in wind engineering, described [here](https://en.wikipedia.org/wiki/Wind_gradient) (for a more official reference, it is also mentioned in the book by Siegfried Heier (2014), Grid Integration of Wind Energy Conversion Systems (publisher: Wiley))

I am willing to take up the task of making the PR, etc., but first, I would like to know your thoughts on which expression should be added (maybe both?) or if there is some other equation you think is more relevant.

### PR DESCRIPTION
**Pull request by IoannisSifnaios**
- [X] Closes #2118 
 - [X] I am familiar with the [contributing guidelines](https://pvlib-python.readthedocs.io/en/latest/contributing.html)
 - [X] Tests added
 - [X] Updates entries in [`docs/sphinx/source/reference`](https://github.com/pvlib/pvlib-python/blob/main/docs/sphinx/source/reference) for API changes.
 - [x] Adds description and name entries in the appropriate "what's new" file in [`docs/sphinx/source/whatsnew`](https://github.com/pvlib/pvlib-python/tree/main/docs/sphinx/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
 - [X] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
 - [x] Pull request is nearly complete and ready for detailed review.
 - [x] Maintainer: Appropriate GitHub Labels (including `remote-data`) and Milestone are assigned to the Pull Request and linked Issue.

The wind speed at different heights is calculated using the Hellmann power law expression ([source 1](https://link-springer-com.proxy.findit.cvt.dk/book/10.1007/3-540-70949-5), [source 2](https://en.wikipedia.org/wiki/Wind_gradient)). The Sandia method (discussed in #2118) is a subcase of this model, so it was added in the notes as a possible coefficient combination that could be used by the user.

### CODE REVIEW

File: pvlib/windspeed.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    height of 30 ft (9.144 m) and the following coefficients were recommended:
+    :math:`h_{ref} = 9.144` [m], :math:`a = 0.219` [-], and
+    :math:`windspeed_{href}` is the wind speed at a height of 9.144 [m].
+
+    It should be noted that the equation retuns a value of NaN if the
+    calculated wind speed is negative or a complex number.
+
+    Warning
+    -------
+    Module temperature functions often require wind speeds at a heigt of 10 m
```
Comment (RDaxini) [inline pvlib/windspeed.py:111]:
```suggestion
    Module temperature functions often require wind speeds at a height of 10 m
```

File: pvlib/windspeed.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+        If both ``exponent`` nor a ``surface_type`` is given. These parameters
+        are mutually exclusive.
+
+    KeyError
+        If the specified ``surface_type`` is invalid.
+
+    Notes
+    -----
+    The equation for calculating the wind speed at a height of :math:`h` is
+    given by the following power law equation [1]_, [2]_:
```
Comment (RDaxini) [inline pvlib/windspeed.py:75]: I wouldn't have written the comma but I might be wrong. I think the comma would be necessary if they are in the same bracket (but I don't know if we can do that here?)
```suggestion
    given by the following power law equation [1]_[2]_:
```

File: pvlib/windspeed.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+    .. math::
+       :label: wind speed
+
+        windspeed_h = windspeed_{href} \cdot \left( \frac{h}{h_{ref}} \right)^a
```
Comment (RDaxini) [inline pvlib/windspeed.py:80]: Could we use a symbol for wind speed? I think it would look neater and more readable. For example, W, U, V would be reasonable I reckon. Or you could go pick something funky from the Greek alphabet if you like
```suggestion
        W_h = W_{href} \cdot \left( \frac{h}{h_{ref}} \right)^a
```

File: pvlib/windspeed.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+    +-----------+--------------------+------------------+------------------+
+    | Stable    | 0.27               | 0.34             | 0.60             |
+    +-----------+--------------------+------------------+------------------+
+
+    In a report by SANDIA [3]_, this equation was experimentally tested for a
+    height of 30 ft (9.144 m) and the following coefficients were recommended:
+    :math:`h_{ref} = 9.144` [m], :math:`a = 0.219` [-], and
+    :math:`windspeed_{href}` is the wind speed at a height of 9.144 [m].
+
+    It should be noted that the equation retuns a value of NaN if the
```
Comment (RDaxini) [inline pvlib/windspeed.py:106]: I don't know if it is better, or makes no difference, to write ``np.nan`` rather than NaN in this situation. Corrected a minor typo in this line anyway though.
```suggestion
    It should be noted that the equation returns a value of NaN if the
```

File: pvlib/windspeed.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+    | Stability | Open water surface | Flat, open coast | Cities, villages |
+    +===========+====================+==================+==================+
+    | Unstable  | 0.06               | 0.10             | 0.27             |
+    +-----------+--------------------+------------------+------------------+
+    | Neutral   | 0.11               | 0.16             | 0.40             |
+    +-----------+--------------------+------------------+------------------+
+    | Stable    | 0.27               | 0.34             | 0.60             |
+    +-----------+--------------------+------------------+------------------+
+
+    In a report by SANDIA [3]_, this equation was experimentally tested for a
```
Comment (RDaxini) [inline pvlib/windspeed.py:101]:
```suggestion
    In a report by Sandia [3]_, this equation was experimentally tested for a
```

**Review by RDaxini — CHANGES_REQUESTED** [review]
Nice work @IoannisSifnaios. Just had a quick read and picked up on a few small things for now

**Comment by cwhanse** [comment]
Reactions rather than a review. Questions for anyone looking at this PR:

1. Naming. Do we need a `wind` (prefer that to `windspeed`) code module? For now, the only use is in temperature functions so this new function could be placed in `pvlib.temperature`. Are there other wind-related functions that may be relevant? I don't think pvlib is the place to put something like a flow simulator to estimate dynamic forces on trackers, for example.

2. Function name. Should be shorter. Does not need to repeat the input to output quantities. So may be as short as `pvlib.wind.hellmann`? That may be too terse.

**Comment by RDaxini** [comment]
I am not knowledgeable enough on the wind/PV subject to comment in detail on 1., but just my two cents on 2. at least...

> 2. Function name. Should be shorter. Does not need to repeat the input to output quantities. So may be as short as `pvlib.wind.hellmann`? That may be too terse.

I am +1 for shortening the function name. I think `pvlib.wind.hellmann` works, but `pvlib.wind.windspeed_hellman` might be clearer, especially if there are a range (not all wind speed) wind-related functions in `pvlib.wind`. That'd be similar to some other function formats, e.g. `pvlib.spectrum.spectral_factor_sapm` --> general subject, then output_model as the function name.

Or if everything in `pvlib.wind` is for calculating wind speed, then going back to a `windspeed` module name with a `pvlib.windspeed.hellmann` function seems okay.

File: pvlib/tests/test_windspeed.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+@pytest.fixture
+def heights_desired():
+    return np.array([5, 5, 5, 10, 20, 30])
+
+
+@pytest.fixture
+def wind_speeds_calc(times):
+    return pd.Series([np.nan, np.nan, 5.067393209486324, 7.2178684911195905,
+                      10.565167835216586, 12.817653329393977], index=times)
```
Comment (echedey-ls) [inline pvlib/tests/test_windspeed.py:55]: I don't want to nitpick too much the tests since that's where we can only taste freedom, but two points:

1. If you need to test scalar inputs, I think it's just better to take the zero or last indexed values of each premise/expected series. Either with the `.iloc` method on a series object, or with the usual `[0]`|`[-1]` on the numpy view. You know, you already are checking that the values are correct with respect to the expected behaviour with the numpy and series tests with `assert_allclose`, and that's perfectly fine. There's usually no need to duplicate data in python.
2. And lastly, I think I mislead you in the test patterns _that I prefer_ in some previous PR when I told you about some example fixture. I just took the simpler one I found for the sake of simplifying without looking too much at its role in the code. Generally, I would group all the fixtures into one:

```python
@pytest.fixture
def wind_speeds_at_heighth_hellman():
    data = pd.DataFrame(
        index=pd.date_range(start="2015-01-01 00:00", end="2015-01-01 05:00", freq="1h"),
        columns=["wind_ref", "height_ref", "height_desired", "wind_desired"],  # or wind_expected
        data=[
            (10, -2, 5, np.nan),
            (-10, 2, 5, np.nan),
            (5, 4, 5, 5.067393209486324),
            (7, 6, 10, 7.2178684911195905),
            (10, 8, 20, 10.565167835216586),
            (12, 10, 30, 12.817653329393977)
        ]
    )
    return data
```

Then you can also take into account the surface type. Maybe one for each entry with the help of another column or a general value your fixture returns. Remember returning a tuple is possible, you can unpack it inside the test function like:

```python
@pytest.fixture
def foo_bar()
    return "foo", "bar"  # returns a tuple with two strings

def test_foo(foo_bar):
    word1, word2 = foo_bar
    assert word1 == "foo"
```

And if you use a dataframe in its own form, you can unpack it like you would do to a dictionary into keyword args:

```python
@pytest.fixture
def foo_bar():
    return pd.DataFrame(data=[["foo", "bar"], ["echedey", "Luis"]], columns=["a","b"])  # returns a dataframe with two columns

def func_under_test(a, b):
    return a + b

def test_foo(foo_bar):
    result = func_under_test(**foo_bar)  # unpacks like a dictionary into series, IIRC
    assert (result == pd.Series(["foobar", "echedeyLuis"])).all()
```

Let me insist in that this last point is just to share that there are other patterns/tools to make tests, that may make them a bit more readable, and, most importantly, keep data nearby. I only request changes regarding 1. If you need anything else, clarifying or whatever, you know how to DM me.

**Review by echedey-ls — COMMENTED** [review]
Nice approach, and sorry for the big points I've written down below. W.r.t. the model, does the height of interest need to be lower than the reference height? At least on wikipedia:

imagen

Btw, a point that may also interest @RDaxini, maybe other maintainers, is reviewing the docs log. It shows `warning`'s, `error`'s and `critical`'s if there are any (well, there's always at least two warnings regarding setuptools replacing distutils and 2 I added unintentionally in a recent pr😞 ).

For example:
1. go to this PR page or the official released one.
2. in the version picker, select `builds`
3. Find the last docs built for PR number, e.g. 2124
4. Expand the last log, the one with `python -m sphinx -T -b html ...`
5. Go find all the warnings, errors or criticals. Hint: use Ctrl+F, only works with expanded log.

File: pvlib/windspeed.py [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+    'stable_air_above_flat_open_coast': 0.40,
+    'unstable_air_above_human_inhabited_areas': 0.27,
+    'neutral_air_above_human_inhabited_areas': 0.34,
+    'stable_air_above_human_inhabited_areas': 0.60,
+}
+
+
+def wind_speed_at_height_hellmann(wind_speed_measured, height_measured,
+                                  height_desired, exponent=None,
+                                  surface_type=None):
```
[the remark this answers is not in the document]
Comment (AdamRJensen) [reply pvlib/windspeed.py:23]: I could get behind that given that the function could also be used for modeled data (reanalysis)

**Comment by IoannisSifnaios** [comment]
> Nice approach, and sorry for the big points I've written down below. W.r.t. the model, does the height of interest need to be lower than the reference height? At least on wikipedia:
> 
> imagen

It is correct that Wikipedia mentions that constraint with the height; however, I can't find where this constraint has come from... the formula is reported in other places without the constraint (which I guess also makes sense since the expression still gives reasonable values...). Any other thoughts on this?

File: pvlib/tests/test_windspeed.py (hunk 6 above)
Comment (IoannisSifnaios) [reply pvlib/tests/test_windspeed.py]: Thank you, @echedey-ls; this was very informative! I tried to implement these suggestions in the tests I wrote.

File: pvlib/windspeed.py [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+    speed to be 10 m: :py:func:`~pvlib.temperature.sapm_cell`,
+    :py:func:`~pvlib.temperature.sapm_module`, and
+    :py:func:`~pvlib.temperature.generic_linear`, whereas the
+    :py:func:`~pvlib.temperature.fuentes` model requires wind speed at 9.144 m.
+
+    Additionally, the heat loss coefficients of some models have been developed
+    for wind speed measurements at 10 m (e.g.,
+    :py:func:`~pvlib.temperature.pvsyst_cell`,
+    :py:func:`~pvlib.temperature.faiman`, and
+    :py:func:`~pvlib.temperature.faiman_rad`).
```
[the remark this answers is not in the document]
Comment (IoannisSifnaios) [reply pvlib/windspeed.py]: Thoughts on removing this from this function and adding it somewhere else @AdamRJensen and @kandersolar?

**Comment by IoannisSifnaios** [comment]
I don't really get this error: 
image

I am testing for a negative height (that would return a complex wind speed), which eventually returns a NaN value. Am I misunderstanding something?

**Comment by cwhanse** [comment]
`isinstance` is checking the type of `wind_speed`, which is either `Series` or `ndarray`. Its not checking the type of elements of `wind_speed`.

Rather than checking for and replacing complex values, I think I would use np.where on `height_desired`

File: pvlib/windspeed.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (AdamRJensen) [reply pvlib/windspeed.py]: I think having a dedicated page would be nice in the future, but to avoid dragging this further out I say let's keep it as is, and then we can always move it.

**Comment by AdamRJensen** [comment]
> @echedey-ls it is correct that Wikipedia mentions that constraint with the height; however, I can't find where this constraint has come from... the formula is reported in other places without the constraint (which I guess also makes sense since the expression still gives reasonable values...). Any other thoughts on this?

Given that the reference we use doesn't state any constraint I think we shouldn't either if we can't find a solid reference for that. Also, the equation is reversible, so it produces consistent results estimating wind speeds at higher and lower height.

File: pvlib/atmosphere.py [hunk 9]
```diff
@@ -533,3 +534,165 @@ def angstrom_alpha(aod1, lambda1, aod2, lambda2):
...
+       :label: wind speed
+
+        U_{w,h} = U_{w,ref} \cdot \left( \frac{h}{h_{ref}} \right)^a
+
+    where :math:`h` [m] is the height at which we would like to calculate the
+    wind speed, :math:`h_{ref}` [m] is the reference height at which the wind
+    speed is known, and :math:`U_{w,h}` [m/s] and :math:`U_{w,ref}`
+    [m/s] are the corresponding wind speeds at these heights. :math:`a` is a
+    value that depends on the surface type. Some values found in the literature
+    [1]_ for :math:`a` are the following:
```
Comment (AdamRJensen) [inline pvlib/atmosphere.py:619]:
```suggestion
        WS_{h} = WS_{ref} \cdot \left( \frac{h}{h_{ref}} \right)^a

    where :math:`h` [m] is the height at which we would like to calculate the
    wind speed, :math:`h_{ref}` [m] is the reference height at which the wind
    speed is known, and :math:`WS_{h}` [m/s] and :math:`WS_{ref}`
    [m/s] are the corresponding wind speeds at these heights. The exponent
    :math:`a` [unitless] depends on the surface type. Some values found in the
    literature [1]_ for :math:`a` are:
```

File: pvlib/atmosphere.py [hunk 10]
```diff
@@ -533,3 +534,165 @@ def angstrom_alpha(aod1, lambda1, aod2, lambda2):
...
+       +-----------+--------------------+------------------+------------------+
+       | Neutral   | 0.11               | 0.16             | 0.40             |
+       +-----------+--------------------+------------------+------------------+
+       | Stable    | 0.27               | 0.34             | 0.60             |
+       +-----------+--------------------+------------------+------------------+
+
+    In a report by Sandia [3]_, this equation was experimentally tested for a
+    height of 30 ft (9.144 m) and the following coefficients were recommended:
+    :math:`h_{ref} = 9.144` [m], :math:`a = 0.219` [-], and
+    :math:`windspeed_{href}` is the wind speed at a height of 9.144 [m].
```
Comment (AdamRJensen) [inline pvlib/atmosphere.py:636]:
```suggestion
    In a report by Sandia [3]_, the equation was experimentally tested for a
    height of 30 ft (:math:`h_{ref} = 9.144` [m]) and a coefficient of
    :math:`a = 0.219` [-] was recommended.
```

File: pvlib/atmosphere.py [hunk 11]
```diff
@@ -533,3 +534,165 @@ def angstrom_alpha(aod1, lambda1, aod2, lambda2):
...
+
+    It should be noted that the equation returns a value of NaN if the
+    calculated wind speed is negative or a complex number.
```
Comment (AdamRJensen) [inline pvlib/atmosphere.py:639]:
```suggestion
    It should be noted that the equation returns a value of NaN if the
    reference heights or wind speed are negative.
```

**Review by AdamRJensen — APPROVED** [review]
I think we can remove all mentions of units as the equation only deals with ratios.

File: pvlib/atmosphere.py (hunk 11 above)
Comment (cwhanse) [reply pvlib/atmosphere.py:639]: > if the reference heights or wind speed are negative

If that's the intent, then the code should test `reference_height` and `wind_speed_reference` rather than `wind_speed`.

File: pvlib/atmosphere.py [hunk 12]
```diff
@@ -533,3 +534,165 @@ def angstrom_alpha(aod1, lambda1, aod2, lambda2):
...
+
+    The model is based on the power law equation by Hellmann [1]_ [2]_.
+
+    Parameters
+    ----------
+    wind_speed_reference : numeric
+        Measured wind speed. [m/s]
+
+    height_reference : float
+        The height at which the wind speed is measured. [m]
```
Comment (cwhanse) [inline pvlib/atmosphere.py:567]:
```suggestion
        The height above ground at which the wind speed is measured. [m]
```

File: pvlib/atmosphere.py [hunk 13]
```diff
@@ -533,3 +534,165 @@ def angstrom_alpha(aod1, lambda1, aod2, lambda2):
...
+
+    height_desired : float
+        The height at which the wind speed will be estimated. [m]
```
Comment (cwhanse) [inline pvlib/atmosphere.py:570]:
```suggestion
        The height above ground at which the wind speed will be estimated. [m]
```

File: pvlib/atmosphere.py (hunk 9 above)
Comment (cwhanse) [reply pvlib/atmosphere.py:619]: OK to not specify units of m/s for `wind_speed_reference`, but a comment that `wind_speed` has the same units as `wind_speed_reference` would be useful.

File: pvlib/atmosphere.py (hunk 10 above)
Comment (wholmgren) [reply pvlib/atmosphere.py:636]: Can we add context on the conditions for which Sandia recommends these coefficients? Was this from their test site in Albuquerque (maybe)? The middle of a forest in Kentucky (probably not)?

File: pvlib/atmosphere.py [hunk 14]
```diff
@@ -533,3 +534,165 @@ def angstrom_alpha(aod1, lambda1, aod2, lambda2):
...
+
+    In a report by Sandia [3]_, this equation was experimentally tested for a
+    height of 30 ft (9.144 m) and the following coefficients were recommended:
+    :math:`h_{ref} = 9.144` [m], :math:`a = 0.219` [-], and
+    :math:`windspeed_{href}` is the wind speed at a height of 9.144 [m].
+
+    It should be noted that the equation returns a value of NaN if the
+    calculated wind speed is negative or a complex number.
+
+    Warning
```
Comment (wholmgren) [inline pvlib/atmosphere.py:641]: I don't think a warning is relevant here. A warning should be about issues when using this function - not issues when using other functions. I would put this material at the top of the notes section.

File: pvlib/atmosphere.py [hunk 15]
```diff
@@ -533,3 +534,165 @@ def angstrom_alpha(aod1, lambda1, aod2, lambda2):
...
+        raise ValueError(
+            "Either a 'surface_type' or an 'exponent' parameter must be given")
+
+    wind_speed = wind_speed_reference * (
+        (height_desired / height_reference) ** exponent)
+
+    # if the provided height is negative the calculated wind speed is complex
+    # so a NaN value is returned
+    if isinstance(wind_speed, complex):
+        wind_speed = np.nan
```
[the remark this answers is not in the document]
Comment (wholmgren) [reply pvlib/atmosphere.py:690]: might as well combine the condition with the test below:

`wind_speed = np.where(np.iscomplex(wind_speed) | wind_speed < 0, np.nan, wind_speed)`

**Review by adriesse — CHANGES_REQUESTED** [review]

File: pvlib/atmosphere.py (hunk 14 above)
Comment (adriesse) [reply pvlib/atmosphere.py:641]: I had a look at the report and it is _very_ limited.  The authors fit 268 pairs of wind speed measurements with values > 2.2 m/s @ 30 ft. which were obtained over the course of 6 weeks.  It also does _not_ _recommend_ a parameter value.

>These results pertain to a single set of data obtained at a single site, so
we cannot claim that they apply elsewhere. Their consistency with the Mechanical
EngineerJ,s Handbook, though, is encouraging. If a model of this form is used,
for other locations, the model should be verified by similarly collecting and
analyzing wind data.

File: docs/sphinx/source/whatsnew/v0.11.1.rst [hunk 16]
```diff
@@ -17,6 +17,9 @@ Enhancements
   :py:func:`pvlib.spectrum.spectral_factor_firstsolar`.
   (:issue:`2086`, :pull:`2100`)
 
+* Added function for calculating wind speed at different heights,
+  :py:func:`pvlib.atmosphere.windspeed_hellmann`.
```
Comment (cwhanse) [inline docs/sphinx/source/whatsnew/v0.11.1.rst:21]:
```suggestion
  :py:func:`pvlib.atmosphere.windspeed_power_law`.
```

**Review by cwhanse — APPROVED** [review]

**Comment by AdamRJensen** [comment]
I've merged the function as the reviewer comments have been addressed (a note was added about Albuquerque location and the function was renamed).

Thanks @IoannisSifnaios 🥳
